common/src/reserved.rs

```diff
@@ -22,55 +22,122 @@ pub struct ReservedState {
+                    .as_ref()
+                    .map_or_else(|| member.name.as_str(), |name| name.as_str());
+                let public_key = self.query_public_key(&name.to_string()).ok_or_else(|| {
+                    format!("the public key of {name} is not found in the reserved state.")
+                })?;
+                Ok((public_key, member.consensus_voting_power))
+            })
+            .try_fold(
+                HashMap::new(),
+                |mut map, result: Result<(crypto::PublicKey, u64), String>| {
+                    let (public_key, voting_power) = result?;
+                    map.entry(public_key)
+                        .and_modify(|v| *v += voting_power)
+                        .or_insert(voting_power);
+                    Ok::<HashMap<crypto::PublicKey, u64>, String>(map)
+                },
+            )?
+            .into_iter()
+            .collect::<Vec<_>>();
+
+        Ok(governance_set)
     }
 
-    pub fn apply_delegate(&mut self, _tx: &TxDelegate) -> Result<Self, String> {
-        unimplemented!()
+    pub fn apply_delegate(&mut self, tx: &TxDelegate) -> Result<Self, String> {
+        let data = DelegationTransactionData {
+            delegator: tx.data.delegator.clone(),
+            delegatee: tx.data.delegatee.clone(),
+            governance: tx.data.governance,
+            block_height: tx.data.block_height,
+            timestamp: tx.data.timestamp,
+        };
+        if tx.proof.verify(&data).is_err() {
+            return Err("delegation proof verification failed".to_string());
+        }
+        let delegator_name = match self.query_name(&tx.data.delegator.clone()) {
+            Some(name) => name,
+            None => return Result::Err("delegator does not exist by name".to_string()),
+        };
+        let delegatee_name = match self.query_name(&tx.data.delegatee.clone()) {
+            Some(name) => name,
+            None => return Result::Err("delegatee does not exist by name".to_string()),
+        };
+        for delegator in &mut self.members {
+            if delegator.name == delegator_name {
+                if tx.data.governance {
+                    delegator.governance_delegatee = Some(delegatee_name.clone());
+                    delegator.consensus_delegatee = Some(delegatee_name.clone());
+                } else {
+                    delegator.consensus_delegatee = Some(delegatee_name.clone());
+                }
+            }
+        }
+        Ok(self.clone())
     }
 
-    pub fn apply_undelegate(&mut self, _tx: &TxUndelegate) -> Result<Self, String> {
-        unimplemented!()
+    pub fn apply_undelegate(&mut self, tx: &TxUndelegate) -> Result<Self, String> {
+        let data = UndelegationTransactionData {
+            delegator: tx.data.delegator.clone(),
+            block_height: tx.data.block_height,
+            timestamp: tx.data.timestamp,
+        };
+        match tx.proof.verify(&data).is_err() {
```

Comment:
ditto

Reply:
would make the change.

Reply:
`tx.proof.verify(&data)` doesn't have a `ok_or_else.` I would rather use the change like:
```rust
if tx.proof.verify(&data).ok().is_none() {
            return Err("undelegation proof verification failed".to_string());
        }
```

Reply:
btw what is the reason for not using match?

Reply:
Ah then try using `map_err` instead.